src/Microsoft.FeatureManagement/ConfigurationFeatureDefinitionProvider.cs

```diff
@@ -83,8 +84,39 @@ public async IAsyncEnumerable<FeatureDefinition> GetAllFeatureDefinitionsAsync()
 
         private FeatureDefinition ReadFeatureDefinition(string featureName)
         {
-            IConfigurationSection configuration = GetFeatureDefinitionSections()
-                                                    .FirstOrDefault(section => section.Key.Equals(featureName, StringComparison.OrdinalIgnoreCase));
+            IConfigurationSection configuration;
+
+            if (!featureName.Contains(ConfigurationPath.KeyDelimiter))
+            {
+                configuration = GetFeatureDefinitionSections()
+                    .FirstOrDefault(section => section.Key.Equals(featureName, StringComparison.OrdinalIgnoreCase));
+            }
+            else
+            {
+                //
+                // Feature names with configuration path delimiters require traversing children sections for resolution
+                IEnumerable<string> sectionNames = featureName.Split(SectionSeparator, StringSplitOptions.None);
+
+                IEnumerable<IConfigurationSection> sections;
+
+                IConfigurationSection section = null;
+
+                foreach (string sectionName in sectionNames)
+                {
+                    sections = section == null ?
+                        GetFeatureDefinitionSections() :
+                        section.GetChildren();
+
+                    section = sections.FirstOrDefault(section => section.Key.Equals(sectionName, StringComparison.OrdinalIgnoreCase));
+
+                    if (section == null)
+                    {
+                        break;
+                    }
+                }
+
+                configuration = section;
```

Comment:
If `featureName` is `Feature:With:Colons`, after the foreach loop ends, we'll set `configuration` to the section named `Colons` . Then we call `ReadFeatureDefinition(configuration)`, and that method will return the `FeatureDefinition` object where `FeatureDefinition.Name` will be set to `Colons` instead of `Feature:With:Colons` ([code](https://github.com/microsoft/FeatureManagement-Dotnet/blob/8d1774ff0521dbbf0f1f35551cc2d7115d1f65e9/src/Microsoft.FeatureManagement/ConfigurationFeatureDefinitionProvider.cs#L172)).

Reply:
Thanks. This made me realize a bigger problem which is how this works with `GetFeatureDefinitionAsync`. We may not be able to handle that

Reply:
GetFeatureDefinitionAsync can't detect features that have `:`. That's a glaring hole in supporting `:`. We may just want to block `:` because of that.